File: 8-string-to-integer-atoi/8-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        Integer res = null;
        Boolean isNeg = null;
        
        int idx = 0;
        for(; idx < s.length() && s.charAt(idx) == ' '; idx++);
        
        for(; idx < s.length(); idx++) {
            char c = s.charAt(idx);
            
            if (isNeg == null && res == null && c == '-') {
                isNeg = true;
                continue;
            }
            
            if (isNeg == null && res == null && c == '+') {
                isNeg = false;
                continue;
            }
            
            if (!Character.isDigit(c)) {
                break;
            }
            
            if (res == null) {
                res = 0;
            }
            
            if (res > Integer.MAX_VALUE / 10 || res == Integer.MAX_VALUE / 10 && c >= '7') {
                return Integer.MAX_VALUE;
            }
            
            if (res < Integer.MIN_VALUE / 10 || res == Integer.MIN_VALUE / 10 && c >= '8') {
                return Integer.MIN_VALUE;
            }
            
            res = res * 10 + ((res >= 0)? Character.getNumericValue(c) : -Character.getNumericValue(c));
            
            if (res != 0 && isNeg != null && isNeg == true) {
                res = -res;
                isNeg = false;
            }
        }
        
        return (res == null)? 0 : res;
    }
}

File: 8-string-to-integer-atoi/8-string-to-integer-atoi.scala
object Solution {
  def myAtoi(s: String): Int = {
    var cur = 0
    var res = 0

    while (cur < s.length && s.charAt(cur) == ' ') {
      cur += 1
    }

    var isNeg = false
    if (cur < s.length && (s.charAt(cur) == '+' || s.charAt(cur) == '-')) {
      isNeg = s.charAt(cur) == '-'
      cur += 1
    }

    var end = false
    while (!end && cur < s.length) {
      if (!s.charAt(cur).isDigit) { end = true }
      else {
        val v = s.charAt(cur).getNumericValue
        if (res > 0 && isNeg) res *= -1

        if (isNeg && (res < (Int.MinValue) / 10 || (res == Int.MinValue / 10 && v >= 8))) {
          res = Int.MinValue
          end = true
        } else if (!isNeg && (res > Int.MaxValue / 10 || (res == Int.MaxValue / 10 && v >= 7))) {
          res = Int.MaxValue
          end = true
        } else {
          if (isNeg) res = res * 10 - v
          else res = res * 10 + v
        }
        cur += 1
      }
    }

    res
  }
}